Fix coverage

d-meiser opened the pull request:
I took a stab at implementing the fix for the generation of gcov/lcov data discussed in #6. This pull request takes the route of using `exit()` instead of `_exit()` if we're building with coverage enabled.

Note that if #5 was resolved (I believe @matthargett  has a patch for that already) then an alternative solution would be to run the test cases without forking. I think that solution might be preferrable because users of cgreen can use the same build of the library with and without gcov.

I can add automatic coverage analysis using coveralls as soon as PR #1 (or some form of it) is merged.

gardenia commented:
@d-meiser - I could be missing the point but I'm unclear on the value of building cgreen (the library) with coverage flags.  or not wrt. issue #6 anyhow.  in issue #6 I'm conceptually trying to qualify the coverage of the code I'm unit testing rather than the coverage of cgreens core.

It looks here like what your change would accomplish would be to allow us to qualify the coverage of the cgreen tests against the cgreen library.  which may in itself be a useful thing to do but is distinct from issue #6 which is about qualifying the coverage of code I'm testing (not cgreen itself).

I'm not qubbling with the general idea of qualifying the coverage of the cgreen tests on the cgreen library.  I'm just giving further context on this pull request wrt. issue #6 in case it wasn't clear originally.

thoni56 commented:
@gardenia that's a very good and important clarification. I was also confused about this I think. Agree that we should keep the two issues separate. Let #6 be about user valuable coverage, and we can keep #7 for the internal need of coveraging cgreen itself (which is not as important).

d-meiser commented:
@gardenia thanks for your comment. I had the two issues -- enabling coverage for the user of cgreen and enabling coverage for cgreen developers -- conflated in my mind. Because of that this patch was not as clear as it should have been.

The primary purpose of this PR is to allow *users* of cgreen to obtain coverage information. I agree this is more important. The key modification is the call of `__gcov_flush()` before `_exit()`ing. In order to do that I had to link against the gcov library. Adding the `-ftest-coverage -fprofile-arcs` is not needed for users of cgreen but enables coverage for cgreen *developers*. So this last point is were I lumped the two objectives together and things got confusing.

I think this approach has some merit because it gives you both uncluttered test output (because you're still using `_exit()` to terminate the child processes) as well as coverage information. I believe the requirement of linking against gcov is not too bad because a user of cgreen will  have to link against it anyways if they wish to get coverage information for their code. But I'm not married to this solution and see the appeal in not having to worry about the coverage stuff at the time the cgreen library is built.

I could  change the CMakeLists.txt  so the `-ftest-coverage -fprofile-arcs` compiler flags are only added if we're building `WITH_UNIT_TESTS`. What do you think?

gardenia commented:
assuming linking to gcov and adding __gcov_flush() solves the coverage prolem outlined in #6 then I think that is a viable option.  my main concern is the optional nature of that solution when it comes to building the library.  if the default is to not build with gcov and __gcov_flush() then effectively I need a custom build of cgreen for coverage to work properly.  if our goal is to have cgreen available as a prebuilt package (as I believe it should be) then it is an extra hoop for people to jump through to get fairly normal behavior.  if we were to making linking to gcov the default link behavior then would that be considered a gratituous dependency for building cgreen for people not interested in coverage or for people who do not have gcov readily available on their systems?

the alternative to the above is to go back to the old exit vs _exit solution and have some runtime means to control which one is used (environment variable or perhaps cgreen_set_child_exit_behavior() type global function).  the benefit of this solution is that it does not require a custom build of cgreen.

@matthargett  - have you any feelings on the above options?

wrt. the -ftest-coverage -fprofile-arcs flags being set for the cgreen core: my opinion is that this should be decoupled from the above so your suggestion of a separate define to control that makes sense to me.

d-meiser commented:
@gardenia I agree with you it might be inconvenient or surprising for users of cgreen to have to link against gcov. Perhaps the best, least cluttered solution would be to just scrap the whole __gcov_flush() business and give users the option of terminating test processes via exit. I think controlling this via an environment variable would be best because I imagine that there are many circumstances where one would want to switch between the two quickly, without code modification and rebuilding.

I can rework this pull request to just provide test coverage for the cgreen library itself. This will work properly as soon as the exit vs _exit mechanism is implemented.

gardenia commented:
@d-meiser sounds good to me.

d-meiser commented:
Hi all,

I modified this PR according to our discussion:
- I renamed the cmake option for controlling the gcov stuff to `INTERNAL_WITH_GCOV`
- The option has been marked as advanced in the cmake configuration
- I added code to .travis.yml to post coverage information to coveralls.

Looks like I still need a bit of debugging on travis. If I can't resolve these issues shortly I'll move it to a different branch and report back here when it's sorted out.

d-meiser commented:
I've been able to reproduce the clang failures seen in the travis builds on a local machine. The segfault occurs after main():

```
Running "libcgreen_runner_tests" (11 tests)...
Completed "Runner": 28 passes, 0 failures, 0 exceptions.
Completed "libcgreen_runner_tests": 28 passes, 0 failures, 0 exceptions.

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7bb4e40 in ?? ()
(gdb) where
#0  0x00007ffff7bb4e40 in ?? ()
#1  0x00007ffff7df3143 in llvm_writeout_files () from /scr_ivyamd/dmeiser/cgreen/cgreen-build-clang/src/libcgreen.so.1
#2  0x0000003b940361bd in __cxa_finalize () from /lib64/libc.so.6
#3  0x00007ffff7de31b6 in __do_global_dtors_aux () from /scr_ivyamd/dmeiser/cgreen/cgreen-build-clang/src/libcgreen.so.1
#4  0x0000000000000000 in ?? ()
```

Other projects have seen similar problems with llvm + coverage on travis, see for instance this bug report:

https://llvm.org/bugs/show_bug.cgi?id=20530

I'll research a bit more to see if there are any solutions. If this turns out to be very difficult or time consuming I propose that we get coverage information from one of the gcc builds.

d-meiser commented:
@thoni56 I've cleaned up the pull request and as far as I'm concerned it's ready to be merged. Could you review it, please. You can see an example of the coverage reports at the following location:

https://coveralls.io/github/d-meiser/cgreen

The url for the badge in the readme points to where the coverage data should be posted for the cgreen-dev github repo. You'll have to sign in to coveralls and enable the repo to make this work.

d-meiser commented on src/posix_runner_platform.c:
```diff
@@ -80,7 +81,11 @@ void run_specified_test_if_child(TestSuite *suite, TestReporter *reporter){
 }
 
 static void stop(void) {
+#ifdef INTERNAL_WITH_GCOV
```
Note that this can be streamlined once #5 is addressed.

thoni56 commented:
@d-meiser do we have to do anything in addition to enabling the repo to get coveralls? I enabled it and then restarted the build but that wasn't picked up. Maybe restart build doesn't cut it, another build is required?

d-meiser commented:
I think this might be because there was no new commit associated with the travis job that generated the coverage data. coveralls seems to get confused by that. For instance on this page https://coveralls.io/jobs/8279187 it reports `First build on master at ?%`. The next commit to the repo should fix this.

Looking at the tail of the travis log it appears that the coverage data was generated successfully and sent over to coveralls:

```
Overall coverage rate:
  lines......: 82.0% (1976 of 2409 lines)
  functions..: 81.3% (331 of 407 functions)
  branches...: 62.2% (339 of 545 branches)
#<Net::HTTPOK 200 OK readbody=true>
{"message":"Job #26.1","url":"https://coveralls.io/jobs/8279196"}
```

And coveralls received the data alright (see for example https://coveralls.io/files/1189688391). It just doesn't generate a proper report.

d-meiser commented:
@thoni56 I noticed that your latest commits to master still haven't updated the coverage status displayed by the coveralls badge. Also, the coverage check is still shown as pending in https://github.com/cgreen-devs/cgreen/pull/13 . I did a bit of research and it appears that coveralls has issues from time to time with their queue processor being overwhelmed. Probably best to give it some more time to see if the problem disappears.

thoni56 commented:
Ok, thanks for the heads-up. I was waiting for coverage before addressing #2 (to see it go down and then up again, maybe ;-).
